File: python/src/aoc/year2025/day1.py
from aoc.util import load_input, load_example


def part1_internal(p, signum, delta):
    return p == 0


def part2_internal(p, signum, delta):
    return (-(signum * p) % 100 + delta) // 100


def solve(lines, f):
    p, total = 50, 0
    for line in lines:
        signum = -1 if line[0] == "L" else 1
        delta = int(line[1:])
        p = (p + signum * delta) % 100
        total += f(p, signum, delta)
    return total


def part1(lines):
    """
    >>> part1(load_example(__file__, "1"))
    3
    """
    return solve(lines, part1_internal)


def part2(lines):
    """
    >>> part2(load_example(__file__, "1"))
    6
    """
    return solve(lines, part2_internal)


if __name__ == "__main__":
    data = load_input(__file__, 2025, "1")
    print(part1(data))
    print(part2(data))

File: rust/src/aoc/year2025/day1.rs
pub fn yd() -> (usize, usize) {
    (2025, 1)
}

struct Dial {
    position: isize,
    total: usize,
}

impl Dial {
    fn part1(&self, delta: isize) -> Self {
        let p1 = (self.position + delta) % 100;
        let total1 = self.total + (p1 == 0) as usize;
        Dial {
            position: p1,
            total: total1,
        }
    }

    fn part2(&self, delta: isize) -> Self {
        let p1 = (self.position + delta) % 100;
        let s = (delta.signum() * self.position).rem_euclid(100) + delta.abs();
        let total1 = self.total + (s / 100) as usize;
        Dial {
            position: p1,
            total: total1,
        }
    }
}

fn solve<F>(data: &str, f: F) -> usize
where
    F: Fn(Dial, isize) -> Dial,
{
    let data = data.replace("L", "-").replace("R", "+");
    let mut dial = Dial {
        position: 50,
        total: 0,
    };
    for line in data.lines() {
        let delta = line.parse::<isize>().unwrap();
        dial = f(dial, delta)
    }
    dial.total
}

pub fn part1(data: &str) -> usize {
    solve(data, |dial, delta| dial.part1(delta))
}

pub fn part2(data: &str) -> usize {
    solve(data, |dial, delta| dial.part2(delta))
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
";

    #[test]
    fn tests_part1() {
        assert_eq!(part1(INPUT), 3);
    }

    #[test]
    fn tests_part2() {
        assert_eq!(part2(INPUT), 6);
    }
}
